llvm/lib/Target/X86/X86ISelLowering.cpp

```diff
@@ -7254,6 +7255,10 @@ static SDValue lowerBuildVectorAsBroadcast(BuildVectorSDNode *BVOp,
     EVT CVT = Ld.getValueType();
     assert(!CVT.isVector() && "Must not broadcast a vector type");
 
+    // 512 bit vpbroadcastw is only available with AVX512BW
+    if (ScalarSize == 16 && IsGT256 && !Subtarget.hasBWI())
+      return SDValue();
```

Comment:
Isn't X86DAGToDAGISel::PreprocessISelDAG() is supposed to handle this?

Reply:
The problem only exists for `v32f16` which is not handled by `PreprocessISelDAG`. Maybe we need add `v32f16` support there.

Reply:
Sounds good. Will add support there.

Reply:
Done with some refactoring. PTAL.